clients/hadoopfs/src/main/java/io/lakefs/LakeFSFileSystem.java

```diff
@@ -223,59 +303,130 @@ public String getScheme() {
      */
     private ObjectLocation pathToObjectLocation(Path path) {
-        if (!path.isAbsolute()) {
-            path = new Path(this.workingDirectory, path);
-        }
-
-        URI uri = path.toUri();
-        if (uri.getScheme() != null && uri.getPath().isEmpty()) {
-            return null;
-        }
-
-        ObjectLocation loc = new ObjectLocation();
-        loc.setRepository(uri.getHost());
-        // extract ref and rest of the path after removing the '/' prefix
-        String s = uri.getPath();
-        if (s.startsWith(SEPARATOR)) {
-            s = s.substring(1);
-        }
-        int i = s.indexOf(SEPARATOR);
-        if (i == -1) {
-            loc.setRef(s);
-        } else {
-            loc.setRef(s.substring(0, i));
-            loc.setPath(s.substring(i+1));
-        }
-        return loc;
+	if (!path.isAbsolute()) {
+	    path = new Path(this.workingDirectory, path);
+	}
+
+	URI uri = path.toUri();
+	if (uri.getScheme() != null && uri.getPath().isEmpty()) {
+	    return null;
+	}
+
+	ObjectLocation loc = new ObjectLocation();
+	loc.setRepository(uri.getHost());
+	// extract ref and rest of the path after removing the '/' prefix
+	String s = trimLeadingSlash(uri.getPath());
+	int i = s.indexOf(SEPARATOR);
+	if (i == -1) {
+	    loc.setRef(s);
+	} else {
+	    loc.setRef(s.substring(0, i));
+	    loc.setPath(s.substring(i+1));
+	}
+	return loc;
+    }
+
+    private static String trimLeadingSlash(String s) {
+	if (s.startsWith(SEPARATOR)) {
+	    return s.substring(1);
+	}
+	return s;
     }
 
     private static class ObjectLocation {
-        private String repository;
-        private String ref;
-        private String path;
-        
-        public String getRepository() {
-            return repository;
-        }
-
-        public void setRepository(String repository) {
-            this.repository = repository;
-        }
-
-        public String getRef() {
-            return ref;
-        }
-
-        public void setRef(String ref) {
-            this.ref = ref;
-        }
-
-        public String getPath() {
-            return path;
-        }
-
-        public void setPath(String path) {
-            this.path = path;
-        }
+	private String repository;
+	private String ref;
+	private String path;
+
+	public String getRepository() {
+	    return repository;
+	}
+
+	public void setRepository(String repository) {
+	    this.repository = repository;
+	}
+
+	public String getRef() {
+	    return ref;
+	}
+
+	public void setRef(String ref) {
+	    this.ref = ref;
+	}
+
+	public String getPath() {
+	    return path;
+	}
+
+	public void setPath(String path) {
+	    this.path = path;
+	}
+    }
+
+    @Override
+    public boolean exists(Path f) throws IOException {
+	return false;
+    }
+
+    /**
+     * Wraps a FSDataOutputStream to link file on staging when done writing
+     */
+    static private class LinkOnCloseOutputStream extends java.io.OutputStream {
```

Comment:
you probably don't need the "java.io" prefix here

Reply:
Ha!  So [the Java Language Specification section 7](https://docs.oracle.com/javase/specs/jls/se10/html/jls-7.html) says that Java auto-imports only _some_ classes, specifically `java.lang.*`.  Instead, I added an explicit import for `OutputStream`.  We still have fewer than 40 import statements, not nearly enough for a Java class...